src/rust/grapl-observe/src/statsd_formatter.rs

```diff
@@ -0,0 +1,199 @@
+use crate::metric_error::MetricError;
+use crate::metric_error::MetricError::{MetricInvalidCharacterError, MetricInvalidSampleRateError};
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::fmt::Write;
+
+lazy_static! {
+    static ref INVALID_CHARS: Regex = Regex::new("[|#,=:]").unwrap();
+}
+
+pub struct TagPair<'a>(&'a str, &'a str);
+
+impl TagPair<'_> {
+    fn write_to_buf(&self, buf: &mut String) -> Result<(), MetricError> {
+        let TagPair(tag_key, tag_value) = self;
+        reject_invalid_chars(tag_key)?;
+        reject_invalid_chars(tag_value)?;
+        Ok(write!(buf, "{}={}", tag_key, tag_value)?)
+    }
+}
+
+fn reject_invalid_chars(s: &str) -> Result<(), MetricError> {
+    let matched = INVALID_CHARS.is_match(s);
+    if matched {
+        Err(MetricInvalidCharacterError())
+    } else {
+        Ok(())
+    }
+}
+
+pub enum MetricType {
+    Gauge,
+    Counter,
+    Millis,
+    Histogram,
+}
+
+impl MetricType {
+    fn statsd_type(&self) -> &'static str {
+        let g: &'static str = "g";
+        let c: &'static str = "c";
+        let ms: &'static str = "ms";
+        let h: &'static str = "h";
+
+        match self {
+            MetricType::Gauge => g,
+            MetricType::Counter => c,
+            MetricType::Millis => ms,
+            MetricType::Histogram => h,
+        }
+    }
+}
+
+pub fn statsd_format(
+    metric_name: &str,
+    value: f64,
+    metric_type: MetricType,
+    sample_rate: Option<f64>, // TODO: change to an `impl Into<f64>`
```

Comment:
Is this TODO still a thing?

Reply:
Also you probably want Into<Option<f64>>